### PR TITLE
http3 transport android

### PR DESCRIPTION
**Pull request by alexis-kochetov**

### CODE REVIEW

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    private IExecutorService Executor { get; }
+
+    public CronetMessageHandler(IExecutorService executor)
+    {
+        Executor = executor;
+        _cronetEngine = new CronetEngine.Builder(MauiApplication.Current)
+            .EnableBrotli(true)
+            .EnableHttp2(true)
+            .EnableQuic(true)
+            .AddQuicHint("dev.actual.chat", 443, 443)
```
Comment (FrolYo) [inline src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:23]: what about local?

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        _cronetEngine.Dispose();
+    }
+
+
+    private class CronetCallback : UrlRequest.Callback
+    {
+        private const int MaxRedirectCount = 3;
+        private static readonly RecyclableMemoryStreamManager MemoryStreamManager = new ();
+
+        private readonly TaskCompletionSource<HttpResponseMessage> _result;
```
Comment (FrolYo) [inline src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:72]: not sure about maui, but don't we TaskSource?

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        public override void OnFailed(UrlRequest p0, UrlResponseInfo p1, CronetException p2)
+        {
+            _responseBodyChannel?.DisposeSilently();
+            _responseBody?.DisposeSilently();
+            _responseBodyChannel = null;
+            _responseBody = null;
+            _result.SetException(p2);
+        }
+
+        public override void OnRedirectReceived(UrlRequest p0, UrlResponseInfo p1, string p2)
```
Comment (FrolYo) [inline src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:99]: why such var names?

Comment (FrolYo) [reply src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:99]: Is it a copy/paste?

File: src/dotnet/App.Maui/Services/NativeHttpClientFactory.Android.cs [hunk 4]
```diff
@@ -1,19 +1,22 @@
...
 {
     private partial HttpMessageHandler? CreatePlatformMessageHandler()
-        => new AndroidMessageHandler {
-            SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls13,
-            AutomaticDecompression = DecompressionMethods.All,
-            UseCookies = true,
-            MaxConnectionsPerServer = 200,
-        };
+        => new CronetMessageHandler(Services.GetRequiredService<IExecutorService>());
+    // => new AndroidMessageHandler {
```
Comment (FrolYo) [inline src/dotnet/App.Maui/Services/NativeHttpClientFactory.Android.cs:14]: remove?

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs (hunk 1 above)
Comment (alexis-kochetov) [reply src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:23]: Nginx doesn't support Quic yet

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs (hunk 2 above)
Comment (alexis-kochetov) [reply src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:72]: Will try to switch...

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs (hunk 3 above)
Comment (alexis-kochetov) [reply src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:99]: those are generated param names from Java classes. Should we rename them?

File: src/dotnet/App.Maui/Services/NativeHttpClientFactory.Android.cs (hunk 4 above)
Comment (alexis-kochetov) [reply src/dotnet/App.Maui/Services/NativeHttpClientFactory.Android.cs:14]: Sure

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs (hunk 2 above)
Comment (alexyakunin) [reply src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:72]: Yep :)

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs (hunk 3 above)
Comment (alexyakunin) [reply src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:99]: IMO it's right to keep the generated names if they are the same as in Java source code.

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        }
+
+        public override void OnReadCompleted(UrlRequest p0, UrlResponseInfo p1, ByteBuffer p2)
+        {
+            if (CancellationToken.IsCancellationRequested) {
+                p0.Cancel();
+                return;
+            }
+            p2.Flip();
+            try
```
Comment (alexyakunin) [inline src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:134]: Bracket should be on the same line inside method body.

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            {
+                _responseBodyChannel ??= Java.Nio.Channels.Channels.NewWritableChannel(_responseBody)!;
+                _responseBodyChannel.Write(p2);
+                p2.Clear();
+                p0.Read(p2);
+            }
+            catch (Exception e)
```
Comment (alexyakunin) [inline src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:141]: Same here.

File: src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                    ret.Content.Headers.Add(key,values);
+                else
+                    ret.Headers.Add(key, values);
+            // we are waiting for transmit completion - but we can try to return Response as soon as possible with uncompleted StreamContent
+            _result.TrySetResult(ret);
+        }
+
+        public override void OnCanceled(UrlRequest request, UrlResponseInfo info)
+        {
+            _responseBodyChannel?.DisposeSilently();
```
Comment (alexyakunin) [inline src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs:182]: Prob. makes sense to move all the disposal code to a single helper method (seems it's repeating in a few places).

**Review by alexyakunin — CHANGES_REQUESTED** [review]

**Review by FrolYo — APPROVED** [review]

**Pull request merged**